sky/utils/command_runner.py

```diff
@@ -301,18 +301,23 @@ def rsync(
             rsync_command.append(
                 RSYNC_EXCLUDE_OPTION.format(str(resolved_source / GIT_EXCLUDE)))
 
+        # rsync doesn't support '~' in a quoted target path. need to expand it.
+        full_source_str = str(resolved_source)
+        if resolved_source.is_dir():
+            full_source_str = os.path.join(full_source_str, '')
+
         ssh_options = ' '.join(
             ssh_options_list(self.ssh_private_key, self.ssh_control_name))
         rsync_command.append(f'-e "ssh {ssh_options}"')
         if up:
             rsync_command.extend([
-                source,
-                f'{self.ssh_user}@{self.ip}:{target}',
+                f'{full_source_str!r}',
```

Comment:
Can we add a comment for why we need the quote for the source and target?

Reply:
Added!